staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -72,10 +74,12 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
 		client := clientValue.Load().(*clientv3.Client)
 		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 		defer cancel()
-		if _, err := client.Cluster.MemberList(ctx); err != nil {
-			return fmt.Errorf("error listing etcd members: %v", err)
+		// See https://github.com/etcd-io/etcd/blob/master/etcdctl/ctlv3/command/ep_command.go#L118
+		_, err := client.Get(ctx, "health")
```

Comment:
under what circumstances will this return unhealthy?

Reply:
According to etcd's guarantees this will fail when the instance it is connected to fails to contact a majority of etcd servers and ask for the data. Practically this means it will succeed as long as the etcd instance it is connected to will return or accept data.

Reply:
I think this makes sense since (as I understand it) a 200 from the etcd health endpoint should establish that we have quorum (@jpbetz to confirm though, he knows way better than I).

Reply:
etcd 3.3+ servers: https://github.com/etcd-io/etcd/blob/release-3.3/etcdserver/api/etcdhttp/metrics.go#L94-L123
etcd 3.2 server: https://github.com/etcd-io/etcd/blob/release-3.2/etcdserver/api/etcdhttp/base.go#L82-L103
etcd 3.1 server: https://github.com/etcd-io/etcd/blob/release-3.1/etcdserver/api/v2http/client.go#L369

All ultimately make a quorum read to ensure a healthy etcd cluster before reporting healthy and will return a http 503 Service Unavailable response if not.

Reply:
I just talked this over with @jpbetz and this doesn't actually do what we initially thought it was doing. This does not actually hit the etcd health endpoint, it actually just makes a query to etcd to get the value for the key 'health'. From a readability perspective, this is really easy code to misunderstand.

However, making a read against etcd here is not a bad idea since it, as a side effect, establishes quorum. @jpbetz suggested making a read against an empty range rather than against the key 'health' since we will avoid hitting a key that may be actually be in use (and potentially have a large value).

Reply:
The old version didn't, but since I removed the permission denied handler (K8s should have access otherwise the health check is no good in case of a permission problem) this is now an issue, you're right. Using the configured prefix is probably best.

Reply:
I've now fixed the health check to request something under the prefix to make sure we do not generate spurious errors when running with strict ACLs.